[SPARK-11131] [core] Fix race in worker registration protocol.

vanzin opened the pull request:
Because the registration RPC was not really an RPC, but a bunch of
disconnected messages, it was possible for other messages to be
sent before the reply to the registration arrived, and that would
confuse the Worker. Especially in local-cluster mode, the worker was
succeptible to receiving an executor request before it received a
message from the master saying registration succeeded.

On top of the above, the change also fixes a ClassCastException when
the registration fails, which also affects the executor registration
protocol. Because the `ask` is issued with a specific return type,
if the error message (of a different type) was returned instead, the
code would just die with an exception. This is fixed by having a common
base trait for these reply messages.

Commit:
[SPARK-11131] [core] Fix race in worker registration protocol.

Because the registration RPC was not really an RPC, but a bunch of
disconnected messages, it was possible for other messages to be
sent before the reply to the registration arrived, and that would
confuse the Worker. Especially in local-cluster mode, the worker was
succeptible to receiving an executor request before it received a
message from the master saying registration succeeded.

On top of the above, the change also fixes a ClassCastException when
the registration fails, which also affects the executor registration
protocol. Because the `ask` was issued with a specific return type,
if the error message (of a different type) was returned instead, the
code will just die with an exception. This is fixed by having a common
base trait for these reply messages.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43800 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43800/consoleFull) for   PR 9138 at commit [`cb3f972`](https://github.com/apache/spark/commit/cb3f972b3e95d70edf16f2b6f04d8b3c48365590).

SparkQA commented:
[Test build #43800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43800/console) for   PR 9138 at commit [`cb3f972`](https://github.com/apache/spark/commit/cb3f972b3e95d70edf16f2b6f04d8b3c48365590).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RegisterWorkerFailed(message: String) extends DeployMessage with RegisterWorkerResponse`
  * `  case class RegisteredExecutor() extends CoarseGrainedClusterMessage`
  * `case class HyperLogLogPlusPlus(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43800/
Test FAILed.

vanzin commented:
pyspark failure, sigh. Is there a bug tracking these flaky tests? retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43808 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43808/consoleFull) for   PR 9138 at commit [`cb3f972`](https://github.com/apache/spark/commit/cb3f972b3e95d70edf16f2b6f04d8b3c48365590).

SparkQA commented:
[Test build #43808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43808/console) for   PR 9138 at commit [`cb3f972`](https://github.com/apache/spark/commit/cb3f972b3e95d70edf16f2b6f04d8b3c48365590).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RegisterWorkerFailed(message: String) extends DeployMessage with RegisterWorkerResponse`
  * `  case class RegisteredExecutor() extends CoarseGrainedClusterMessage`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43808/
Test PASSed.

vanzin commented:
/cc @andrewor14 @zsxwing I think you're the people most familiar with this code.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -387,6 +362,31 @@ private[deploy] class Master(
+      logInfo("Registering worker %s:%d with %d cores, %s RAM".format(
+        workerHost, workerPort, cores, Utils.megabytesToString(memory)))
+      if (state == RecoveryState.STANDBY) {
+        // ignore, don't send response
```
this is kind of scary. We should always send a response otherwise we might get random future timeout exception

vanzin replied:
Perhaps, but not really related to the problem at hand.

andrewor14 replied:
what I'm saying is in general in `receiveAndReply` you should always reply because that's what the caller expects. Otherwise it's really confusing when we get future timeouts cause they're hard to debug.

vanzin replied:
I agree with you, I'm just saying that's a bug in itself, and I'd rather not make that as part of this, because that might affect other parts of the code where the workers retry connections to different masters.

andrewor14 replied:
wait, actually that's a problem. Right now we have a `send` and an `ask`. The `send` actually won't be received by anyone because we only handle this in `receiveAndReply`

vanzin replied:
I might have missed a send. But the lack of a reply here is not necessarily a bug, and is definitely not related to this one.

vanzin replied:
Actually, nevermind. I see what you mean. I'm introducing the need to reply by switching to `ask`. Let me take a look...

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -36,9 +36,13 @@ private[spark] object CoarseGrainedClusterMessages {
-  case object RegisteredExecutor extends CoarseGrainedClusterMessage
+  sealed trait RegisterExecutorResponse
+
+  case class RegisteredExecutor() extends CoarseGrainedClusterMessage
```
why change this from case object?

vanzin replied:
because it needs to extend the trait, and it didn't work when I did not make that change. (I also have a change built on top of this one where I'll have to add a parameter to this message.)

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -341,25 +351,36 @@ private[deploy] class Worker(
-      }, 0, HEARTBEAT_MILLIS, TimeUnit.MILLISECONDS)
-      if (CLEANUP_ENABLED) {
-        logInfo(s"Worker cleanup enabled; old application directories will be deleted in: $workDir")
+  private def handleRegisterResponse(msg: RegisterWorkerResponse): Unit = synchronized {
```
if this is called only from the actor then you don't need synchronized here I don't think

vanzin replied:
it's not called from the actor, it's called from a netty i/o thread.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -341,25 +351,36 @@ private[deploy] class Worker(
+        logInfo("Successfully registered with master " + masterRef.address.toSparkURL)
+        registered = true
+        changeMaster(masterRef, masterWebUiUrl)
         forwordMessageScheduler.scheduleAtFixedRate(new Runnable {
```
not your code, but I just noticed this should be forwArd

andrewor14 commented:
@vanzin just so I understand, it's these two lines that are racing, right?

https://github.com/apache/spark/blob/8ac71d62d976bbfd0159cac6816dd8fa580ae1cb/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L249

https://github.com/apache/spark/blob/8ac71d62d976bbfd0159cac6816dd8fa580ae1cb/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L703

vanzin commented:
@andrewor14 correct, that's what causes the race.

Commit:
Fix other `send` call site.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43876 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43876/consoleFull) for   PR 9138 at commit [`ff2d3e2`](https://github.com/apache/spark/commit/ff2d3e25494b288db08a544b752cac998d4f7178).

SparkQA commented:
[Test build #43876 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43876/console) for   PR 9138 at commit [`ff2d3e2`](https://github.com/apache/spark/commit/ff2d3e25494b288db08a544b752cac998d4f7178).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RegisterWorkerFailed(message: String) extends DeployMessage with RegisterWorkerResponse`
  * `  case class RegisteredExecutor() extends CoarseGrainedClusterMessage`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43876/
Test PASSed.

Commit:
Explicitly reply even when master is in standby.

Also go back to using case objects since they seem to work; I'll make
the change to case classes where needed in the corresponding PR (not
yet sent).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43884 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43884/consoleFull) for   PR 9138 at commit [`2d35024`](https://github.com/apache/spark/commit/2d350249bbf40baa7677b150e4481c675784e9f2).

SparkQA commented:
[Test build #43884 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43884/console) for   PR 9138 at commit [`2d35024`](https://github.com/apache/spark/commit/2d350249bbf40baa7677b150e4481c675784e9f2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RegisterWorkerFailed(message: String) extends DeployMessage with RegisterWorkerResponse`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43884/
Test PASSed.

andrewor14 commented:
LGTM I'm merging this into master thanks @vanzin

zsxwing commented:
@vanzin just found an issue about this change. Now if the master receives `RegisterWorker`, it won't use the `workerRef` to send the reply. So there is no connection from `Master` to the server in `Worker`. If the `Worker` is killed now, `Master` only observes some client is lost, but the address is just a client address in Worker and won't match the Worker address. So `Master` cannot remove this dead `Worker` at once. However, this Worker will be removed in 60 seconds because of no heartbeat.

See the log here: 
https://www.mail-archive.com/dev@spark.apache.org/msg12332.html

andrewor14 commented:
To echo @vanzin on SPARK-12267, the cause is not this PR but #9210.
